File: app/src/main/java/br/com/fclug/financialaid/LoginActivity.java
package br.com.fclug.financialaid;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.HttpURLConnection;

import br.com.fclug.financialaid.server.ApiRequest;
import br.com.fclug.financialaid.server.ServerUtils;

public class LoginActivity extends AppCompatActivity {

    private EditText mUsername;
    private EditText mPassword;
    private Button mSignIn;
    private Button mSignUp;
    private TextView mSkipLogin;

    private String usernameString;
    private String passwordString;
    private JSONObject requestArgs;

    private ProgressDialog mProgressDialog;
    private AlertDialog mAlertDialog;

    private TextWatcher mWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            usernameString = mUsername.getText().toString();
            passwordString = mPassword.getText().toString();
            mSignIn.setEnabled(!usernameString.trim().isEmpty() && !passwordString.trim().isEmpty());
        }

        @Override
        public void afterTextChanged(Editable s) {

        }
    };

    private ApiRequest.RequestCallback mLoginCallback = new ApiRequest.RequestCallback() {
        @Override
        public void onSuccess(JSONObject response) throws JSONException {
            String token = null;
            String name = null;
            if (response != null) {
                token = response.getString("token");
                name = response.getString("name");
            }
            mProgressDialog.dismiss();
            // Creating user login session
            SessionManager.createLoginSession(LoginActivity.this, usernameString, name, passwordString, token);
            SessionManager.sendFbTokenToServer(LoginActivity.this);

            // Start MainActivity
            Intent i = new Intent(getApplicationContext(), MainActivity.class);
            startActivity(i);
            finish();
        }

        @Override
        public void onFailure(int code) {
            mProgressDialog.dismiss();
            switch (code) {
                case HttpURLConnection.HTTP_GATEWAY_TIMEOUT:
                    Toast.makeText(LoginActivity.this, "Server is unavailable", Toast.LENGTH_SHORT).show();
                    break;
                case HttpURLConnection.HTTP_UNAUTHORIZED:
                    mAlertDialog.setMessage("Username and password do not match");
                    mAlertDialog.show();
                    break;
                case HttpURLConnection.HTTP_INTERNAL_ERROR:
                    mAlertDialog.setMessage("Unable to connect to server");
                    mAlertDialog.show();
                    break;
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        mUsername = (EditText) findViewById(R.id.login_username);
        mPassword = (EditText) findViewById(R.id.login_password);
        mSignIn = (Button) findViewById(R.id.login_signin);
        mSignUp = (Button) findViewById(R.id.login_signup);
        mSkipLogin = (TextView) findViewById(R.id.login_skip);

        mUsername.addTextChangedListener(mWatcher);
        mPassword.addTextChangedListener(mWatcher);
        mSignIn.setEnabled(false);


        mProgressDialog = new ProgressDialog(LoginActivity.this);
        mAlertDialog = new AlertDialog.Builder(LoginActivity.this).create();
        mAlertDialog.setTitle("Login failed!");
        mAlertDialog.setButton(DialogInterface.BUTTON_POSITIVE, "OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                mUsername.getText().clear();
                mPassword.getText().clear();
            }
        });

        mSignIn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                mProgressDialog.setTitle("Loading");
                mProgressDialog.setMessage("Authenticating...");
                mProgressDialog.show();

                requestArgs = new JSONObject();
                try {
                    requestArgs.put("username", usernameString);
                    requestArgs.put("password", passwordString);
                } catch (JSONException e) {
                    e.printStackTrace();
                }

                new ApiRequest(ServerUtils.METHOD_POST, ServerUtils.ROUTE_LOGIN, requestArgs, mLoginCallback).execute();

            }
        });

        mSignUp.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(LoginActivity.this, RegisterActivity.class);
                startActivity(intent);
            }
        });

        mSkipLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                SessionManager.skipLogin(LoginActivity.this);
            }
        });

    }
}

File: app/src/main/java/br/com/fclug/financialaid/LoginActivity.kt
package br.com.fclug.financialaid

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import br.com.fclug.financialaid.server.*
import br.com.fclug.financialaid.utils.TAG
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_login.*
import org.koin.android.ext.android.inject
import retrofit2.HttpException

class LoginActivity : AppCompatActivity() {

    private var disposable: Disposable? = null
    private val api: ServerApi by inject()

    private val loginWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            login_signin.isEnabled = login_username.text.toString().trim { it <= ' ' }.isNotEmpty()
                    && login_password.text.toString().trim { it <= ' ' }.isNotEmpty()
        }

        override fun afterTextChanged(s: Editable) {}
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        login_username.addTextChangedListener(loginWatcher)
        login_password.addTextChangedListener(loginWatcher)


        val progressDialog = AlertDialog.Builder(this)
                .setView(R.layout.progress_dialog)
                .setTitle(resources.getString(R.string.login_dialog_title))
                .setCancelable(false)
                .create()

        val alertDialog = AlertDialog.Builder(this)
                .setTitle(resources.getString(R.string.login_failed))
                .create()
        alertDialog.setButton(DialogInterface.BUTTON_POSITIVE, "OK") { dialog, which ->
            login_username.text.clear()
            login_password.text.clear()
        }

        val sessionManager: SessionManager by inject()

        login_signin.setOnClickListener {

            val username = login_username.text.toString()
            val password = login_password.text.toString()
            val userLogin = UserLogin(username, password)

            progressDialog.show()
            disposable = api.login(userLogin)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({response ->
                        progressDialog.dismiss()

                        val name = response.result.name
                        val token = response.result.token

                        // Creating user login session
                        sessionManager.createLoginSession(username, name, password, token)
                        sessionManager.sendFbTokenToServer()
                        // Start MainActivity
                        val i = Intent(applicationContext, MainActivity::class.java)
                        startActivity(i)
                        finish()
                    }, {error ->
                        progressDialog.dismiss()

                        if(error is HttpException) {
                            error.response()?.errorBody()?.let {
                                val errorResponse = api
                                        .convertErrorResponseBody<ApiResponseError>(it)
                                errorResponse?.let { response ->
                                    alertDialog.setMessage(response.errors[0].description)
                                    alertDialog.show()
                                }
                            }
                        } else {
                            Toast.makeText(this@LoginActivity, R.string.api_request_failed,
                                    Toast.LENGTH_SHORT).show()
                            Log.d(TAG, error.message)
                            error.printStackTrace()
                        }
                    })

        }

        login_signup.setOnClickListener {
            val intent = Intent(this@LoginActivity, RegisterActivity::class.java)
            startActivity(intent)
        }

        login_skip.setOnClickListener { sessionManager.skipLogin() }
    }
}